### PR TITLE
For getting activations on harmful prompts

### PR DESCRIPTION
**Pull request by revmag**

### CODE REVIEW

File: pipeline/run_pipeline.py [hunk 1]
```diff
@@ -39,186 +93,533 @@ def load_and_sample_datasets(cfg):
...
+        os.path.join(cfg.artifact_path(), "generate_directions/mean_diffs.pt"),
+    )
 
     return mean_diffs
 
-def select_and_save_direction(cfg, model_base, harmful_val, harmless_val, candidate_directions):
+
+def normalize_vector(vector):
+    """Normalize a given vector to have unit length."""
+    norm = torch.norm(vector)
```
Comment (taufeeque9) [inline pipeline/run_pipeline.py]:
```suggestion
    norm = torch.linalg.vector_norm(vector)
```

Use `torch.linalg.vector_norm` here. `torch.norm` is deprecated for good reasons (like confusing matrix norm with vector norm): https://pytorch.org/docs/stable/generated/torch.norm.html

My guess is most likely this is a major bug in the code that is not letting us reproduce the paper's results.

Also add an assert statement to confirm the shape of `vector` is as you expect it to be for this function. Maybe `assert len(vector.shape) == 1`?

File: pipeline/run_pipeline.py [hunk 2]
```diff
@@ -39,186 +93,533 @@ def load_and_sample_datasets(cfg):
...
+    if norm == 0:
```
Comment (taufeeque9) [inline pipeline/run_pipeline.py]:
```suggestion
    if th.isclose(norm, th.tensor(0.))
```

File: pipeline/run_pipeline.py [hunk 3]
```diff
@@ -39,186 +93,533 @@ def load_and_sample_datasets(cfg):
...
+    plt.legend()
+
+    plot_file_path = os.path.join(activations_dir, "dot_products_plot.png")
+    plt.savefig(plot_file_path)
+    plt.close()
+    print(f"Plot saved to {plot_file_path}")
+
+
+def computing_dot_product(
+    data_list,
```
Comment (taufeeque9) [inline pipeline/run_pipeline.py]: It would be good to have types next to the arguments

File: pipeline/run_pipeline.py [hunk 4]
```diff
@@ -39,186 +93,533 @@ def load_and_sample_datasets(cfg):
...
+    refusal_direction,
+    average_dot_products,
+    resultant_dot_products,
+    activations_dir,
+    layers,
+    harmful_train,
+):
+    data_list = [normalize_vector(data) for data in data_list]
```
Comment (taufeeque9) [inline pipeline/run_pipeline.py]: add assert statements to check the shape/size of the input matches your expectation

File: pipeline/run_pipeline.py [hunk 5]
```diff
@@ -39,186 +93,533 @@ def load_and_sample_datasets(cfg):
...
+    data_list = []
+
+    for i in range(len(harmful_train)):
+        mean_activations = generate_activations_for_harmful(
+            model_base,
+            harmful_train[i],
+        )
+        file_path = os.path.join(activations_dir, f"mean_activations_for_{i}_prompt.pt")
+
+        # mean activations has a size of 5*18*d_model -> it takes activations of last 5 positions of prompt
```
Comment (taufeeque9) [inline pipeline/run_pipeline.py:323]: this comment should be an assert statement. Actually check within the code whether `mean_activations` has the shape that you expect.

File: pipeline/run_pipeline.py [hunk 6]
```diff
@@ -39,186 +93,533 @@ def load_and_sample_datasets(cfg):
...
+        data_list.append(mean_activations[4, :, :])
```
Comment (taufeeque9) [inline pipeline/run_pipeline.py]: Index using -1 instead of 4? Its more robust that way

File: pipeline/run_pipeline.py (hunk 3 above)
Comment (taufeeque9) [reply pipeline/run_pipeline.py]: + a better name. `data_list`  doesn't convey anything. `layer_wise_activations` is a better name for example.

File: pipeline/run_pipeline.py [hunk 7]
```diff
@@ -39,186 +93,533 @@ def load_and_sample_datasets(cfg):
...
+    activations_dir,
+    layers,
+    harmful_train,
+):
+    data_list = [normalize_vector(data) for data in data_list]
+
+    """Calculate and save dot products, then generate and save a DataFrame with results."""
+    for i in range(layers):
+        # Calculate average dot products
+        dot_products = sum(
```
Comment (taufeeque9) [inline pipeline/run_pipeline.py]: Always use torch / numpy functions when dealing with tensors instead of native python functions like sum

File: pipeline/run_pipeline.py [hunk 8]
```diff
@@ -39,186 +93,533 @@ def load_and_sample_datasets(cfg):
...
+            for j in range(len(harmful_train))
+        ) / len(harmful_train)
+        average_dot_products.append(dot_products)
+
+        # Calculate resultant dot products
+        resultant_vector = sum(data_list[j][i] for j in range(len(harmful_train)))
+        resultant_vector = normalize_vector(resultant_vector)
+        dot_product = np.dot(
+            resultant_vector.cpu().numpy(), refusal_direction.cpu().numpy()
+        )
```
Comment (taufeeque9) [inline pipeline/run_pipeline.py]:
```suggestion
        dot_product = torch.dot(resultant_vector, refusal_direction)
```
When you already have torch tensors, prioritize using torch function instead of converting it to numpy arrays and using numpy functions.

File: pipeline/run_pipeline.py [hunk 9]
```diff
@@ -39,186 +93,533 @@ def load_and_sample_datasets(cfg):
...
+        resultant_dot_products.append(dot_product)
+
+    print("Average Dot Products:", average_dot_products)
+    print("Resultant Dot Products:", resultant_dot_products)
+
+    # Create and display a DataFrame with results
+    data = {
+        "Layer": list(range(layers)),
+        "Sum along cos(A_i, R)": average_dot_products,
```
Comment (taufeeque9) [inline pipeline/run_pipeline.py]:
```suggestion
        "mean(cos(A_i, R))": average_dot_products,
```

File: pipeline/run_pipeline.py [hunk 10]
```diff
@@ -39,186 +93,533 @@ def load_and_sample_datasets(cfg):
...
+        "Cos(resultant(A_i), R)": resultant_dot_products,
```
Comment (taufeeque9) [inline pipeline/run_pipeline.py]:
```suggestion
        "cos(mean(A_i), R)": resultant_dot_products,
```

File: pipeline/run_pipeline.py [hunk 11]
```diff
@@ -39,186 +93,533 @@ def load_and_sample_datasets(cfg):
...
+    layers, average_dot_products, resultant_dot_products, activations_dir
+):
+    layers=range(layers)
+    plt.figure(figsize=(10, 5))  # Set the figure size as needed
+
+    # Plotting Average Dot Products
+    plt.plot(layers, average_dot_products, label="Average Dot Product", marker="o")
+
+    # Plotting Resultant Dot Products
+    plt.plot(layers, resultant_dot_products, label="Resultant Dot Product", marker="o")
```
Comment (taufeeque9) [inline pipeline/run_pipeline.py]:
```suggestion
    plt.plot(layers, resultant_dot_products, label="similarity(mean(A_i), R)", marker="o")
```

File: pipeline/run_pipeline.py [hunk 12]
```diff
@@ -39,186 +93,533 @@ def load_and_sample_datasets(cfg):
...
+
+
+def plotting_refusal_with_activations(
+    layers, average_dot_products, resultant_dot_products, activations_dir
+):
+    layers=range(layers)
+    plt.figure(figsize=(10, 5))  # Set the figure size as needed
+
+    # Plotting Average Dot Products
+    plt.plot(layers, average_dot_products, label="Average Dot Product", marker="o")
```
Comment (taufeeque9) [inline pipeline/run_pipeline.py]:
```suggestion
    plt.plot(layers, average_dot_products, label="mean(similarity(A_i, R))", marker="o")
```

**Review by taufeeque9 — CHANGES_REQUESTED** [review]
Requested changes. Please resolve all requests and request a review again.

File: pipeline/run_pipeline.py [hunk 13]
```diff
@@ -194,10 +203,16 @@ def generate_and_save_candidate_directions(
     return mean_diffs
 
 
-def normalize_vector(vector):
+def normalize_vector(vector,model_base):
     """Normalize a given vector to have unit length."""
-    norm = torch.norm(vector)
-    if norm == 0:
+    if vector.dim() > 1:
+        assert vector.shape == (model_base.model.config.num_hidden_layers, model_base.model.config.hidden_size), "Shape doesnt match"
```
Comment (taufeeque9) [inline pipeline/run_pipeline.py]: I would recommend not checking for this shape. Your `normalize_vector` function should work for vector/matrix of any sizes. So you can normalize other vectors using this function too.

Just the separate check for `vector.dim()` is good. Your function documentation should mention what happens for multi-dimensional tensor inputs (that they are normalized along the last dim)

File: pipeline/run_pipeline.py [hunk 14]
```diff
@@ -227,41 +242,43 @@ def plotting_refusal_with_activations(
...
+        average_dot_products.append(dot_products.item())
 
         # Calculate resultant dot products
-        resultant_vector = sum(data_list[j][i] for j in range(len(harmful_train)))
-        resultant_vector = normalize_vector(resultant_vector)
-        dot_product = np.dot(
-            resultant_vector.cpu().numpy(), refusal_direction.cpu().numpy()
-        )
-        resultant_dot_products.append(dot_product)
+        resultant_vector = torch.sum(torch.stack([layer_wise_activations[j][i].to(device) for j in range(len(harmful_train))]), dim=0)
```
Comment (taufeeque9) [inline pipeline/run_pipeline.py]:
```suggestion
        resultant_vector = layer_wise_activations[:, i].mean(dim=0)
```
Stacking is very memory-inefficient (it creates new copies of your data in memory, which is very expensive). Use it in rare occasions only.

File: pipeline/run_pipeline.py [hunk 15]
```diff
@@ -227,41 +242,43 @@ def plotting_refusal_with_activations(
...
+    device=refusal_direction.device
     """Calculate and save dot products, then generate and save a DataFrame with results."""
     for i in range(layers):
         # Calculate average dot products
-        dot_products = sum(
-            np.dot(data_list[j][i].cpu().numpy(), refusal_direction.cpu().numpy())
-            for j in range(len(harmful_train))
-        ) / len(harmful_train)
-        average_dot_products.append(dot_products)
+        dot_products = torch.sum(torch.stack([torch.dot(layer_wise_activations[j][i].to(device), refusal_direction) for j in range(len(harmful_train))])) / len(harmful_train)
```
Comment (taufeeque9) [inline pipeline/run_pipeline.py]:
```suggestion
        dot_products = (layer_wise_activations_tensor[:, i] @ refusal_direction).mean()
```
Stacking is very memory-inefficient (it creates new copies of your data in memory, which is very expensive). Use it in rare occasions only.

File: pipeline/run_pipeline.py [hunk 16]
```diff
@@ -227,41 +242,43 @@ def plotting_refusal_with_activations(
...
+) -> None:
+    layer_wise_activations = [normalize_vector(data,model_base) for data in layer_wise_activations]
+
+    #layer_wise_activations is list, converting it to tensors for matching shape
+    layer_wise_activations_tensor = torch.stack(layer_wise_activations)
+    assert layer_wise_activations_tensor.shape == (len(harmful_train), model_base.model.config.num_hidden_layers, model_base.model.config.hidden_size)
+        
+    device=refusal_direction.device
     """Calculate and save dot products, then generate and save a DataFrame with results."""
     for i in range(layers):
```
Comment (taufeeque9) [inline pipeline/run_pipeline.py]: There is no need to go through this for loop. You can use torch function to do everything without any for loop (which is way more efficient as torch functions are written in C++ and they operations in parallel across thousands of gpu-cores). With the below suggestions, you should get an idea on how to remove this for loop also.

File: pipeline/run_pipeline.py [hunk 17]
```diff
@@ -351,36 +369,29 @@ def select_and_save_direction(
     return pos, layer, direction
 
 
-def select_and_save_direction_2(
+def for_refusal_metric_plots(
```
Comment (taufeeque9) [inline pipeline/run_pipeline.py]: You can remove this function completely by moving the makedirs line inside `refusal_metric_plot`?

File: pipeline/run_pipeline.py [hunk 18]
```diff
@@ -480,150 +492,113 @@ def run_pipeline(model_path, refusal_direction=None, activation_prompts=False):
...
-    )
-    evaluate_loss_for_datasets(
-        cfg, model_base, actadd_fwd_pre_hooks, actadd_fwd_hooks, "actadd"
-    )
+        actadd_fwd_pre_hooks, actadd_fwd_hooks = (
+            [
+                (
+                    model_base.model_block_modules[layer],
+                    get_activation_addition_input_pre_hook(
+                        vector=refusal_direction, coeff=-1.0
```
Comment (taufeeque9) [inline pipeline/run_pipeline.py:549]: Can you confirm whether `coeff=-1.0` is correct or not?

**Review by taufeeque9 — CHANGES_REQUESTED** [review]
Requested changes to remove for loop and torch.stack. Please re-request review once done.

File: pipeline/run_pipeline.py (hunk 18 above)
Comment (revmag) [reply pipeline/run_pipeline.py:549]: yes, 
        vector = vector.to(activation)
        activation += coeff * vector
        if isinstance(input, tuple):
            return (activation, *input[1:])
        else:
            return activation
            
    so they are doing 2 things on harmful dataset, 
    1)first ablating using the standard method ( x- (x.r)x)
    2) they are just directly subtracting the refusal direction ( x- r )

File: pipeline/run_pipeline.py (hunk 16 above)
Comment (revmag) [reply pipeline/run_pipeline.py]: Should I remove all for loop then?

Comment (revmag) [reply pipeline/run_pipeline.py]: using einsum?

Comment (revmag) [reply pipeline/run_pipeline.py]: yo facing some issues, here
so to pass it on normalize function, i have to use for loop and the output of that is a list, and to convert the list back to tensor( because einsum accepts only tensor) I have to use torch.stack

So, either I introduce 
1)another variable to store and not use stack / 
2)use for loop and then stack the output

#Method 1
```
normalized_resultant_vectors = torch.zeros_like(resultant_vectors)
    for i in range(resultant_vectors.shape[0]):
        normalized_resultant_vectors[i] = normalize_vector(resultant_vectors[i], model_base)
        
```
#Method 2
```
normalized_resultant_vectors = [normalize_vector(resultant_vectors[i], model_base) for i in range(resultant_vectors.shape[0])]
normalized_resultant_vectors_tensor = torch.stack(normalized_resultant_vectors)
```

Comment (taufeeque9) [reply pipeline/run_pipeline.py]: Einsum is great if you know how to use it. However, you can do most things directly with torch and not require einsum (unless you're doing complicated stuff that torch doesn't provide functions for which is very rare). Just go through some tutorial on how to use pytorch.

```
normalized_resultant_vectors = normalize_vector(resultant_vectors, model_base)
```

You're adding for loops unnecessarily. If you're using a for loop and stack afterwards, it most likely means you're doing it wastefully and you can directly use torch function to do what you want. In this case, `normalize_vector` anyway normalizes along the last dim, so why go through a for loop and stack again when you can do it all in one torch operation?
